### ISSUE TITLE
misc

### ISSUE THREAD

**Issue by Thorin-Oakenpants** [arkenfox/TZP2#4]
things
- [x] viewport scrollbar can = -1 - https://github.com/arkenfox/TZP2/commit/d71d67d50dcdd607c59cc782c3335cc03b4d251c
- [x] speech engines = type error object - https://github.com/arkenfox/TZP2/commit/623d3ae00a6d87abda1be4a1ccd5c62d5edbbbad
- [x] non-RFP: anypointer css says fine but matchmedia says coarse
   - this seems like a bug? creepy reports the same
   - https://github.com/arkenfox/TZP2/commit/4d4f6d577c964a18e95c1608f378560d0519db8e
 - [x] combine any-pointer into a single line ` css1 & css2 | matchmedia1 & matchmedia2`
- [x] RFP notate any-pointer for desktop
   - [x] but check what the RFP patches are actually doing as this was news to me, is it also pointer? etc
- [x] full screen test broken on android - would open full screen and not exit with the element obscuring everything
- [x] android extra tests not displayed (toolbar height, keyboard height etc)
- [x] android new window test returns 0x0, but on a retry or two, returns our values
   - currently code is different for android vs desktop (desktop checks until it errors, android is a one off)
   - `DONE:` expanded first/early android inner to also include outer. This provides all the info needed if it's not protected
   - e.g. 411 x 812 inner vs 414 x 817 = 3px width and 5pixels height for edges
   - e.g. then screen = 414 x 897 = 80px for all chrome (basically should be the toolbar which always displays AFAIK on new urls)
- [x] android: stop spamming click perf with visualViewport info, and stop updating the visualViewport html data: triggered by listeners for scrolling and keyboard height test
- [x] android: expand inner and screen css lower limit to 400
- [x] matchmedia inner/screen was promise.reject when returning `between values`, this meant it would only return the one axis (width or height). Changed to promise.resolve
- [x] fixup perf.now RFP: older devices, cold starts, etc can produce some weird lags between timings, so expand isRFPGood, and relax having both 16.6 + 16.7 to at least either
- [x] element rerun on android causes the page to jump up. The div with all the resizing is offscreen and top -5000px or whatever .. to investigate - moved to #5 
- [x] full screen element when already fullscreen doesn't exit, also allow transition delay
- [x] don't notate netinfo connection object as when rfp green it's at odds with the red default settings for connection navigator, and when rfp red, it's redundant for health purposes
- don't bypass or notate things not protected
   - [x] devicepixelratio (window) - do not bypass, move to own line and only notate only (we record the DPR PoC). -moz-device-pixel-ratio is not protected so also don't notate. don't notate PoC, notations are for health (and I do not need to highlight bypasses, only check patches)
   - [x] dpi
   - [x] ua: checked we no longer bypass the version
- [x] ua: add health failures if the static values change from expected: i.e the whole ua is protected

Actually... any-pointer
- creepy says `@media` = coarse and matchmedia = fine
- tzp2 says `@media` = finie and matchmedia = coarse
- WTF
- to make matters more confusing, TZP 1 says fine + fine
- this is on my touch laptop
- @abrahamjuliot

**Comment by Thorin-Oakenpants** [comment]
creepy
- any-pointer is coarse
![Creep media](https://user-images.githubusercontent.com/16656956/233866584-c25dec65-8a1f-41bd-8dc6-14c4be9510bf.png)

- any-pointer is fine
![Creep matchmedia](https://user-images.githubusercontent.com/16656956/233866654-f12087ce-2d62-45b9-8026-7b3a8684973b.png)

TZP2
- the first two are css then the last  two are matchmedia
![TZP2](https://user-images.githubusercontent.com/16656956/233866663-1b44471c-0846-474c-838b-df6a5043e794.png)

TZP1
![TZP1](https://user-images.githubusercontent.com/16656956/233866682-6e28344a-ffcf-4a95-b907-4f6571b3baec.png)

@abrahamjuliot

**Comment by abrahamjuliot** [comment]
Hm... I wonder if it is unique to touchscreen w/trackpad.

**Comment by Thorin-Oakenpants** [comment]
yeah ... BUT ... why are all three results different ? creepy vs tzp2 = back to front, and tzp1 is different again. I've checked the code in TZP1 v 2 is the same, and AFAICT it is .. so WTF man :)

having a hard time working from my laptop - I prefer my desktops with massive screen real estate, plus I'm lacking a few decent tools on here - am in costa rica with tor project for the week

**Comment by Thorin-Oakenpants** [comment]
OK, tested with a few other people here with laptops - it is because I have a touch screen.

Still doesn't explain whjy tzp2 vs creepy have the values back to front.

**Comment by abrahamjuliot** [comment]
It might be due to the use of invisible iframe on creepy. I need to get rid of that at some point. It's not great for testing top-level results.

**Comment by Thorin-Oakenpants** [comment]
ok, solved why TZP vs TZP2 is back to front

TZP is concatenating the two matchmedia results and the order is pointer, then any-pointer
```
		// pointer
		get_mm("pointer", "#cssP")
		get_mm("any-pointer", "#cssAP")
		let p = display.join(" | ")
		// hover
		display = []
		get_mm("hover", "#cssH")
		get_mm("any-hover", "#cssAH")
		let h = display.join(" | ")
```

TZPs is two separate results and goes any-pointer then pointer

I will make TZP follow suit and retest

**Comment by Thorin-Oakenpants** [comment]
no wait, while I made it easier to check the diffs the problem with TZP1 is

> to make matters more confusing, TZP 1 says fine + fine

**Comment by Thorin-Oakenpants** [comment]
interesting
```js
function get_mm(type, id) {

	// TZP1
	let x=zNS, x2="", f="fine", c="coarse", h="hover", n="none", q=type+":"
	// TZP2
	let value = zNS // TZP2

	try {
		// TZP1
		if (window.matchMedia("("+ q + n +")").matches) x=n
		if (window.matchMedia("("+ q + c +")").matches) x=c
		if (window.matchMedia("("+ q + f +")").matches) x=f
		if (window.matchMedia("("+ q + h +")").matches) x=h

		// TZP2
		if (window.matchMedia("("+ type +":fine").matches) value = "fine"
		if (window.matchMedia("("+ type +":hover)").matches) value = "hover"
		if (window.matchMedia("("+ type +":coarse)").matches) value = "coarse"
		if (window.matchMedia("("+ type +":none)").matches) value = "none"
	} catch(e) {
		x = zErr
	}
	// css value
	x2 = getElementProp(id,"content",":after")

	console.log(type +", "+ id +", tzp1: "+ x +", css: "+ x2 +", tzp2: "+ value)

```

```
any-pointer, #cssAP, tzp1: fine, css: fine, tzp2: coarse
```

**Comment by Thorin-Oakenpants** [comment]
it's the order or checks and no else if's .. digging further. strange that the css doesn't match

**Comment by Thorin-Oakenpants** [comment]
https://www.w3.org/TR/mediaqueries-4/

emphasis mine

> The [any-pointer](https://www.w3.org/TR/mediaqueries-4/#descdef-media-any-pointer) and [any-hover](https://www.w3.org/TR/mediaqueries-4/#descdef-media-any-hover) media features are identical to the [pointer](https://www.w3.org/TR/mediaqueries-4/#descdef-media-pointer) and [hover](https://www.w3.org/TR/mediaqueries-4/#descdef-media-hover) media features, but they correspond to the union of capabilities of all the pointing devices available to the user. **_In the case of any-pointer, more than one of the values can match_**, if different pointing devices have different characteristics.

**Comment by Thorin-Oakenpants** [comment]
aww snap, pointer is the same

**Comment by Thorin-Oakenpants** [comment]
![image](https://user-images.githubusercontent.com/16656956/236605807-1eb7ccb3-85e3-4301-b143-eeddeca61bef.png)

ta-dah

**Comment by Thorin-Oakenpants** [comment]
and with rfp (will add notations)

![image](https://user-images.githubusercontent.com/16656956/236605854-82018462-3e63-4d40-87a8-24bd474b4663.png)

**Comment by Thorin-Oakenpants** [comment]
and on my phone without RFP I get

```
[coarse] any-pointer | coarse | coarse
[fine] any-pointer | coarse | coarse
```

so there you go. Because any-pointer can be both or either that's 3 results possible. `none` only returns if both fine and coarse are not supported. So we need to check order wise (and set in css orderwise) two tests to get both values

@abraham so that explains why all out results were different between TZP1 vs TZP2 vs creepy

**Comment by Thorin-Oakenpants** [comment]
tl;dr: we need to check both states (by altering the order of fine + coarse) to get both values, and combined they make a metric
